### PR TITLE
Task 5: Show/Edit/Delete functions

### PR DESCRIPTION
**Pull request by Luke-kb**
## Description
This PR combines Tasks 5 & 6: adding show/edit/delete methods as well as specs for each method.
some basic pagination styling and a few tweaks to the PR template, dev:populate task and student model validations included.

## Breakdown of Commits
Commit 1 : [ea268a2](https://github.com/Luke-kb/mySMS/commit/ea268a20dd33f5898a02ff7540bf9f89bead6604)
- basic pagination styling. minor changes to PR template
 
Commit 2 : [e911fc8](https://github.com/Luke-kb/mySMS/commit/e911fc8baa80c60131d3b7f1d5e10ea0b9048131) 
- added edit and delete methods

Commit 3: [d9b0479](https://github.com/Luke-kb/mySMS/commit/d9b047918098a58473c5f1e9c27f6d2775d96807)
- updated setup instructions. increased amount of entries to populate db

Commit 4: [6f0c36f](https://github.com/Luke-kb/mySMS/commit/6f0c36f85d589441814e670b8e236115e68a8bdb)
- added byebug and pry-rails

Commit 5: [0417e61](https://github.com/Luke-kb/mySMS/commit/0417e61156ab4a0f5c9881bacf7d62012d3ab2da)
- bundle install gem.lock

Commit 6: [e8e4a57](https://github.com/Luke-kb/mySMS/commit/e8e4a577d056e0efbc8a825929c8ae8852d7c92d)
- added nil exclusion validations to Student model class

Commit 7: [7420cfa](https://github.com/Luke-kb/mySMS/commit/7420cfa300fd6c5a54543a020027e03ba089da5e)
- StudentController specs

#### TODO Checklist
* [x] pagination styling
* [x] added edit / delete methods
* [x] added pry-rails and byebug
* [x] StudentsController specs

### CODE REVIEW

File: app/controllers/students_controller.rb [hunk 1]
```diff
@@ -1,5 +1,32 @@
 class StudentsController < ApplicationController
+   
     def index
         @students = Student.paginate(page: params[:page], per_page: 10)
     end
-end
\ No newline at end of file
+
+    def show
+        @student = Student.find(params[:id])
```
Comment (jbat) [inline app/controllers/students_controller.rb]: @Luke-kb  .. whenever you see yourself using a piece of code repetitively
it normally means you can dry it up by refactoring into another method.
You are loading the student instance variable in 4 different methods.  show, edit, update, destroy

A common practice in controllers is to use a before_filter hook.
ref https://guides.rubyonrails.org/v3.2.13/action_controller_overview.html 
search for before_filter

```
# add this after the class declaration .. 
before_filter :load_student, except: [:index]

# then make a private method at the bottom of the class .. and remove all the other calls to
# @student = Student.find(params[:id])
 
def load_student
  @student = Student.find(params[:id])
end
```

File: spec/controllers/students_controller_spec.rb [hunk 2]
```diff
@@ -0,0 +1,96 @@
+require 'rails_helper'
+
+RSpec.describe StudentsController, type: :controller do
+    #save instance of student in test db
```
Comment (jbat) [inline spec/controllers/students_controller_spec.rb]: comments like this are not really helpfull .. can remove it.
If you want to pass something to the PR review .. as it as a PR comment inline where the code

File: app/assets/stylesheets/pagination.css [hunk 3]
```diff
@@ -0,0 +1,25 @@
+.apple_pagination {
```
Comment (Luke-kb) [inline app/assets/stylesheets/pagination.css:1]: borrowed from [https://mislav.github.io/will_paginate/]

File: app/assets/stylesheets/pagination.css [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+  line-height: 1.4em;
+  -moz-border-radius: 1em;
+  -webkit-border-radius: 1em;
+  border-radius: 1em;
+  text-shadow: rgba(255, 255, 255, 0.8) 1px 1px 1px; }
+.apple_pagination a {
+  text-decoration: none;
+  color: black; }
+.apple_pagination a:hover, .apple_pagination a:focus {
+  text-decoration: underline; }
```
Comment (Luke-kb) [inline app/assets/stylesheets/pagination.css:24]: majority of css borrowed from [will_paginate example css](https://mislav.github.io/will_paginate/)

File: spec/controllers/students_controller_spec.rb [hunk 5]
```diff
@@ -0,0 +1,96 @@
+require 'rails_helper'
+
+RSpec.describe StudentsController, type: :controller do
+    #save instance of student in test db 
+    let!(:student) { create(:student, title: "Mr", first_name: "Harry", middle_name: "James", last_name: "Potter", email: "harry@test.com", gender: "Male", birth_date: "1989-07-23") }
```
Comment (jbat) [inline spec/controllers/students_controller_spec.rb]: @Luke-kb Always create your objects with a purpose in mind.
As a reviewer when I look at this student object, with Mr Harry James Potter .. etc etc
I would expect you to use all those values somewhere below in some context, however that are not used.

So be explicit: .. for this Student object .. who cares what the values are, let the faker gem take care of it.
```
let!(:student) { create(:student) }
```

File: spec/controllers/students_controller_spec.rb [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+        it "assigned requested student to @student" do
+            subject
+            expect(assigns(:student)).to eq student
+        end
+        
+    end
+    
+    describe "PUT #update" do
+        context "with valid attributes" do
+            subject { put :update, id: student, student: attributes_for(:student, first_name: "Larry", last_name: "Potter", email: "harry@test.com", gender: "Male", birth_date: "1989-07-23") }
```
Comment (jbat) [inline spec/controllers/students_controller_spec.rb]: Again .. your only checking the first_name attr below, which is fine.
so here you only need to actually change the first_name: 'Larry' attr, and the rest can be random.
```
subject { put :update, id: student, student: attributes_for(:student, first_name: "Larry") }
```

File: spec/controllers/students_controller_spec.rb (hunk 5 above)
Comment (Luke-kb) [reply spec/controllers/students_controller_spec.rb:5]: Ah I see. So use literal values as per the [recommendation](https://jobready.atlassian.net/wiki/spaces/DEV/pages/43941936/Code+Review+Checklist) only for the values that you intend to test?

File: spec/controllers/students_controller_spec.rb [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+RSpec.describe StudentsController, type: :controller do
+    #save instance of student in test db 
+    let!(:student) { create(:student, title: "Mr", first_name: "Harry", middle_name: "James", last_name: "Potter", email: "harry@test.com", gender: "Male", birth_date: "1989-07-23") }
+    
+    describe "GET #index" do
+        subject { get :index }
+
+        it "has a 200 status code" do
+            subject
+            expect(subject.status).to eq(200)
```
Comment (jbat) [inline spec/controllers/students_controller_spec.rb]: @Luke-kb  here your actually calling subject twice.
If you removed line 11 totally, the result should be still the same
as line 12 will call `subject` and then status on that subject.
Try that .. 
However there is a more common way that you will see the subject pattern being used.
That is with a One-liner syntax
https://relishapp.com/rspec/rspec-core/docs/subject/one-liner-syntax

I suggest you see if you can refactor the other specs below to the oneliner syntax
```
    subject { get :index }
    
    it { is_expected.to be_success }
    it { is_expected.to render_template("students/index") }
    it { expect(assigns(:students)).to eq([student]) }

```

**Review by jbat — CHANGES_REQUESTED** [review]
@Luke-kb have left some comments

**Review by jbat — APPROVED** [review]

**Pull request merged**